# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Как выполнить задание?

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

## Содержание

1. [Цель исследования](#step1)
2. [Подключение](#step2)
3. [Загрузка таблиц](#step23)
3. [Задача 1](#step3)
4. [Задача 2](#step4)
5. [Задача 3](#step5)
6. [Задача 4](#step6)
7. [Задача 5](#step7)

### Цель исследования<a id="step1"></a>

Проанализировать базу данных

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Доступ к базе данных

Следуйте инструкции, приложенной ниже

[instructions_for_connecting_to_the_database.pdf](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/bac02c0e-9489-4458-acff-4cd0ad912b46/instructions_for_connecting_to_the_database.pdf)

### Примечания

- Не забывайте про функции, они значительно облегчат жизнь при выполнении запросов;
- Получите ответы, используя только язык запросов SQL. Pandas допускается только для хранения и вывода результатов.

<a id="step2"></a>

**Импортируем библиотеки**

In [1]:
import pandas as pd
from sqlalchemy import create_engine

**Подключимся к базе данных.**

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

<a id="step23"></a>

**Загрузим таблицу `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [3]:
query = '''

SELECT *
FROM books
LIMIT 5

'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Загрузим таблицу `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [4]:
query = '''

SELECT *
FROM authors
LIMIT 5

'''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Загрузим таблицу `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [5]:
query = '''

SELECT *
FROM ratings
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Загрузим таблицу `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [6]:
query = '''

SELECT *
FROM reviews
LIMIT 5
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Загрузим таблицу `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [7]:
query = '''

SELECT *
FROM publishers
LIMIT 5

'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<a id="step3"></a>

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [8]:
query = '''

SELECT  COUNT(book_id)      
FROM books
WHERE publication_date > '2000-01-01'

'''
pd.io.sql.read_sql(query, con = engine)

count
0    819

**819 книг вышло после 1 января 2000 года.**

<a id="step4"></a>

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [9]:
query = '''

WITH
t AS (SELECT  book_id,
              AVG(rating)
      FROM ratings
      GROUP BY book_id),

v AS (SELECT  book_id,
              COUNT(review_id)
      FROM reviews
      GROUP BY book_id)

SELECT  title,
        avg,
        count

FROM books AS b
LEFT JOIN t on b.book_id = t.book_id
LEFT JOIN v on b.book_id = v.book_id
ORDER BY (avg, count) DESC

'''
pd.io.sql.read_sql(query, con = engine)

title   avg  count
0     The Cat in the Hat and Other Dr. Seuss Favorites  5.00    NaN
1                        A Dirty Job (Grim Reaper  #1)  5.00    4.0
2              School's Out—Forever (Maximum Ride  #2)  5.00    3.0
3         Moneyball: The Art of Winning an Unfair Game  5.00    3.0
4                         Plum Island (John Corey  #1)  5.00    2.0
..                                                 ...   ...    ...
995  The World Is Flat: A Brief History of the Twen...  2.25    3.0
996                                      Drowning Ruth  2.00    3.0
997                                              Junky  2.00    2.0
998                  His Excellency: George Washington  2.00    2.0
999                               Harvesting the Heart  1.50    2.0

[1000 rows x 3 columns]

**В данной таблице к каждой книге представлена средняя оценка и количество обзоров.**

<a id="step5"></a>

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц
— так мы исключим из анализа брошюры

In [10]:
query = '''

SELECT  publisher,
        COUNT(book_id)
                
FROM books AS b
LEFT JOIN publishers AS p on b.publisher_id = p.publisher_id

WHERE num_pages > 50
GROUP BY publisher
ORDER BY COUNT(book_id) DESC
LIMIT 1

'''
pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

**Penguin Books - издательство выпустившее наибольшее число книг толще 50 станиц.** 

<a id="step6"></a>

### Определим автора с самой высокой средней оценкой книг
— учитывайте только книги с 50 и более оценками.

In [11]:
query = '''

WITH
b AS (SELECT  book_id,
              AVG(rating)
      FROM ratings
      GROUP BY book_id
      HAVING COUNT(rating)>=50 
      ORDER BY AVG(rating) DESC),
      
a AS (SELECT  b.book_id,
              b.author_id,
              author
      FROM books AS b
      JOIN authors AS a on b.author_id = a.author_id)

SELECT  author,
        AVG(avg)
        
FROM b
JOIN a on b.book_id = a.book_id
GROUP BY author  
ORDER BY avg DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

**J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой книг**

<a id="step7"></a>

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
query = '''

WITH
t AS (SELECT username,
             COUNT(rating)
      FROM ratings
      GROUP BY username
      HAVING COUNT(rating)>50),

v AS (SELECT  username,
              COUNT(review_id)
      FROM reviews
      GROUP BY username)

SELECT AVG(v.count)
FROM t
JOIN v on t.username = v.username

'''
pd.io.sql.read_sql(query, con = engine)


avg
0  24.333333

**24.33333 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок.**